### ISSUE TITLE
memory leak in bcmtf

### ISSUE THREAD

**Issue by fredRos** [bat/bat#166]
_by email_

I noticed the presence of a memory leak https://github.com/bat/bat/blob/master/models/mtf/BCMTF.cxx#L230
where the `TH1D* hist_uncbandexp` and `hist_uncbandpoisson` are created but not deleted.

``` c++

    TH2D* hist_uncbandexp = new TH2D(Form("UncertaintyBandExpectation_%s_%d", GetSafeName().data(), channelindex), "", hist.GetNbinsX(), &a[0], 1000, minimum, maximum);
    hist_uncbandexp->SetStats(kFALSE);

    TH2D* hist_uncbandpoisson = new TH2D(Form("UncertaintyBandPoisson_%s_%d", GetSafeName().data(), channelindex), "", hist.GetNbinsX(), &a[0], int(maximum - minimum), minimum, maximum);
    hist_uncbandpoisson->SetStats(kFALSE);
```

If one runs just a single fit, it doesn't cause any trouble,
but if one has to do some sensitivity study it might fill up the entire memory at some point...

For the moment I just commented it, but it would be nice to solve the problem (maybe just avoid using pointer?).

**Comment by fredRos** [comment]
After a long pause, I have time to look into these issues again. Before the histograms are added, any old ones are removed in https://github.com/bat/bat/blob/master/models/mtf/BCMTF.cxx#L209

```c++
    // remove old uncertainty histograms if they exist
    if (channel->GetHistUncertaintyBandExpectation()) {
        delete channel->GetHistUncertaintyBandExpectation();
        channel->SetHistUncertaintyBandExpectation(0);
    }
    if (channel->GetHistUncertaintyBandPoisson()) {
        delete channel->GetHistUncertaintyBandPoisson();
        channel->SetHistUncertaintyBandPoisson(0);
    }
...

    TH2D* hist_uncbandexp = new TH2D(Form("UncertaintyBandExpectation_%s_%d", GetSafeName().data(), channelindex), "", hist.GetNbinsX(), &a[0], 1000, minimum, maximum);
    hist_uncbandexp->SetStats(kFALSE);

    TH2D* hist_uncbandpoisson = new TH2D(Form("UncertaintyBandPoisson_%s_%d", GetSafeName().data(), channelindex), "", hist.GetNbinsX(), &a[0], int(maximum - minimum), minimum, maximum);
    hist_uncbandpoisson->SetStats(kFALSE);

    // set histograms
    data->SetHistogram(new TH1D(hist), hist.Integral());
    channel->SetHistUncertaintyBandExpectation(hist_uncbandexp);
    channel->SetHistUncertaintyBandPoisson(hist_uncbandpoisson);

```
so the memory leak can't accumulate, it's limited to the size of these two histograms. The proper place to delete these two histograms finally is the destructor of BCMTFChannel

https://github.com/bat/bat/blob/master/models/mtf/BCMTFChannel.cxx#L35

The necessary code is there but commented out, I'm guessing because some plot with a `TCanvas` would be empty otherwise. I will check this

**Comment by fredRos** [comment]
It turns out the root of all the problems is ROOT! From its docs on creating a [histogram](https://root.cern.ch/doc/master/classTH1.html#ab9c5981066a117bba152690422fd1a9c)

> When an histogram is created, a reference to it is automatically added to the list of in-memory objects for the current file or directory. This default behaviour can be changed by:
h->SetDirectory(0);          for the current histogram h
TH1::AddDirectory(kFALSE);   sets a global switch disabling the reference
>
> When the histogram is deleted, the reference to it is removed from the list of objects in memory. When a file is closed, all histograms in memory associated with this file are automatically deleted.

That's why our `mtf/ensemble` example fails because there is a root file that is opened, so when we delete the histograms we created in the dtor, it may come after the file is closed and root has already deleted the histograms. Insane! This may affect other places as well

**Comment by fredRos** [comment]
This particular issue is fixed in #229 but it may come up in several other places. We have to think of a general strategy, so I'm creating a new issue
